Certificate error with v6 tasks

I try to migrate v5 tasks to v6, but there is an error during the connection to the Octopus Deploy server.

```
2023-03-15T10:39:48.5790955Z ##[section]Starting: Push Package to Octopus
2023-03-15T10:39:48.6089532Z ==============================================================================
2023-03-15T10:39:48.6089809Z Task         : Push Package(s) to Octopus
2023-03-15T10:39:48.6090098Z Description  : Push your NuGet or Zip package to your Octopus Deploy Server. **v6 of this step requires Octopus 2022.3+**
2023-03-15T10:39:48.6090376Z Version      : 6.0.502
2023-03-15T10:39:48.6090540Z Author       : Octopus Deploy
2023-03-15T10:39:48.6090776Z Help         : Version: 6.0.502. [More Information](https://g.octopushq.com/TFS-VSTS)
2023-03-15T10:39:48.6091058Z ==============================================================================
2023-03-15T10:39:48.9160922Z ##[error]"Failed to execute push. unable to get local issuer certificate
Error: unable to get local issuer certificate
    at Function.ae.from (F:\agent3\_work\_tasks\OctopusPush_d05ad9a2-5d9e-4a1c-a887-14034334d6f2\6.0.502\index.js:42:13882)
    at en.<anonymous> (F:\agent3\_work\_tasks\OctopusPush_d05ad9a2-5d9e-4a1c-a887-14034334d6f2\6.0.502\index.js:44:10223)
    at en.emit (node:events:527:28)
    at ClientRequest.Tf.<computed> (F:\agent3\_work\_tasks\OctopusPush_d05ad9a2-5d9e-4a1c-a887-14034334d6f2\6.0.502\index.js:42:738)
    at ClientRequest.emit (node:events:527:28)
    at TLSSocket.socketErrorListener (node:_http_client:454:9)
    at TLSSocket.emit (node:events:527:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
2023-03-15T10:39:48.9252160Z ##[section]Finishing: Push Package to Octopus
```

The certificate used on the Octopus Deploy server is signed by an internal CA.

Currently it is not possible to provide a CA file in Octo tasks, maybe a parameter that contains a path to a CA file on the server will help.

--- a/source/tasks/BuildInformation/BuildInformationV6/buildInformation.ts
+++ b/source/tasks/BuildInformation/BuildInformationV6/buildInformation.ts
@@ -1,24 +1,17 @@
 import { OctoServerConnectionDetails } from "../../Utils/connection";
 import { createCommandFromInputs } from "./inputCommandBuilder";
-import { BuildInformationRepository, Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { BuildInformationRepository, Logger } from "@octopusdeploy/api-client";
 import { TaskWrapper } from "tasks/Utils/taskInput";
 import { getOverwriteMode } from "./overwriteMode";
 import { IVstsHelper } from "./vsts";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
+import { getClient } from "../../Utils/client";
 
 export class BuildInformation {
     constructor(readonly connection: OctoServerConnectionDetails, readonly logger: Logger, readonly task: TaskWrapper, readonly vsts: IVstsHelper) {}
 
     public async run() {
         const command = await createCommandFromInputs(this.logger, this.task, this.vsts);
-
-        const config: ClientConfiguration = {
-            userAgentApp: getUserAgentApp("build-information", "push", 6),
-            instanceURL: this.connection.url,
-            apiKey: this.connection.apiKey,
-            logging: this.logger,
-        };
-        const client = await Client.create(config);
+        const client = await getClient(this.connection, this.logger, "build-information", "push", 6)
 
         const overwriteMode = await getOverwriteMode(this.logger, this.task);
         this.logger.debug?.(`Build Information:\n${JSON.stringify(command, null, 2)}`);

--- a/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV6/release.ts
+++ b/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV6/release.ts
@@ -1,23 +1,14 @@
-import {
-    Client,
-    ClientConfiguration,
-    CreateReleaseCommandV1,
-    Logger,
-    Project,
-    ProjectRepository,
-    Space,
-    SpaceRepository
-} from "@octopusdeploy/api-client";
+import { Client, CreateReleaseCommandV1, Logger, Project, ProjectRepository, Space, SpaceRepository } from "@octopusdeploy/api-client";
 import { OctoServerConnectionDetails } from "../../Utils/connection";
 import { createReleaseFromInputs } from "./createRelease";
 import { createCommandFromInputs } from "./inputCommandBuilder";
 import os from "os";
 import { TaskWrapper } from "tasks/Utils/taskInput";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
 import path from "path";
 import { getVstsEnvironmentVariables } from "../../../tasksLegacy/Utils/environment";
 import { v4 as uuidv4 } from "uuid";
 import * as tasks from "azure-pipelines-task-lib";
+import { getClient } from "../../Utils/client";
 
 export class Release {
     constructor(readonly connection: OctoServerConnectionDetails, readonly task: TaskWrapper, readonly logger: Logger) {}
@@ -25,15 +16,7 @@ export class Release {
     public async run() {
         try {
             const command = createCommandFromInputs(this.logger, this.task);
-
-            const config: ClientConfiguration = {
-                userAgentApp: getUserAgentApp("release", "create", 6),
-                instanceURL: this.connection.url,
-                apiKey: this.connection.apiKey,
-                logging: this.logger,
-            };
-            const client = await Client.create(config);
-
+            const client = await getClient(this.connection, this.logger, "release", "create", 6);
             const version = await createReleaseFromInputs(client, command, this.task, this.logger);
 
             await this.tryCreateSummary(client, command, version);

--- a/source/tasks/Deploy/DeployV6/deploy.ts
+++ b/source/tasks/Deploy/DeployV6/deploy.ts
@@ -1,10 +1,10 @@
-import { Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { Logger } from "@octopusdeploy/api-client";
 import { OctoServerConnectionDetails } from "../../Utils/connection";
 import { createDeploymentFromInputs } from "./createDeployment";
 import { createCommandFromInputs } from "./inputCommandBuilder";
 import os from "os";
 import { TaskWrapper } from "tasks/Utils/taskInput";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
+import { getClient } from "../../Utils/client";
 
 export class Deploy {
     constructor(readonly connection: OctoServerConnectionDetails, readonly task: TaskWrapper, readonly logger: Logger) {}
@@ -12,14 +12,7 @@ export class Deploy {
     public async run() {
         try {
             const command = createCommandFromInputs(this.logger, this.task);
-
-            const config: ClientConfiguration = {
-                userAgentApp: getUserAgentApp("release", "deploy", 6),
-                instanceURL: this.connection.url,
-                apiKey: this.connection.apiKey,
-                logging: this.logger,
-            };
-            const client = await Client.create(config);
+            const client = await getClient(this.connection, this.logger, "release", "deploy", 6);
 
             createDeploymentFromInputs(client, command, this.task, this.logger);
 

--- a/source/tasks/DeployTenant/TenantedDeployV6/deploy.ts
+++ b/source/tasks/DeployTenant/TenantedDeployV6/deploy.ts
@@ -1,10 +1,10 @@
-import { Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { Logger } from "@octopusdeploy/api-client";
 import { OctoServerConnectionDetails } from "../../Utils/connection";
 import { createDeploymentFromInputs } from "./createDeployment";
 import { createCommandFromInputs } from "./inputCommandBuilder";
 import os from "os";
 import { TaskWrapper } from "tasks/Utils/taskInput";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
+import { getClient } from "../../Utils/client";
 
 export class Deploy {
     constructor(readonly connection: OctoServerConnectionDetails, readonly task: TaskWrapper, readonly logger: Logger) {}
@@ -12,14 +12,7 @@ export class Deploy {
     public async run() {
         try {
             const command = createCommandFromInputs(this.logger, this.task);
-
-            const config: ClientConfiguration = {
-                userAgentApp: getUserAgentApp("release", "deploy-tenanted", 6),
-                instanceURL: this.connection.url,
-                apiKey: this.connection.apiKey,
-                logging: this.logger,
-            };
-            const client = await Client.create(config);
+            const client = await getClient(this.connection, this.logger, "release", "deploy-tenanted", 6);
 
             createDeploymentFromInputs(client, command, this.task, this.logger);
 

--- a/source/tasks/Push/PushV6/index.ts
+++ b/source/tasks/Push/PushV6/index.ts
@@ -1,10 +1,10 @@
-import { Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { Logger } from "@octopusdeploy/api-client";
 import * as tasks from "azure-pipelines-task-lib/task";
 import { getDefaultOctopusConnectionDetailsOrThrow } from "../../Utils/connection";
 import { getLineSeparatedItems, getOverwriteModeFromReplaceInput, getRequiredInput } from "../../Utils/inputs";
 import { Push } from "./push";
 import os from "os";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
+import { getClient } from "../../Utils/client";
 
 async function run() {
     try {
@@ -29,15 +29,7 @@ async function run() {
             },
         };
 
-        const config: ClientConfiguration = {
-            userAgentApp: getUserAgentApp("package", "push", 6),
-            instanceURL: connection.url,
-            apiKey: connection.apiKey,
-            logging: logger,
-        };
-
-        const client: Client = await Client.create(config);
-
+        const client = await getClient(connection, logger, "package", "push", 6);
         await new Push(client).run(spaceName, packages, overwriteMode);
     } catch (error: unknown) {
         if (error instanceof Error) {

--- a/source/tasks/RunRunbook/RunRunbookV6/runbookRun.ts
+++ b/source/tasks/RunRunbook/RunRunbookV6/runbookRun.ts
@@ -1,10 +1,10 @@
-import { Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { Logger } from "@octopusdeploy/api-client";
 import { OctoServerConnectionDetails } from "../../Utils/connection";
 import { createRunbookRunFromInputs } from "./runRunbook";
 import { createCommandFromInputs } from "./inputCommandBuilder";
 import os from "os";
 import { TaskWrapper } from "tasks/Utils/taskInput";
-import { getUserAgentApp } from "../../Utils/pluginInformation";
+import { getClient } from "../../Utils/client";
 
 export class RunbookRun {
     constructor(readonly connection: OctoServerConnectionDetails, readonly task: TaskWrapper, readonly logger: Logger) {}
@@ -12,14 +12,7 @@ export class RunbookRun {
     public async run() {
         try {
             const command = createCommandFromInputs(this.logger, this.task);
-
-            const config: ClientConfiguration = {
-                userAgentApp: getUserAgentApp("runbook", "run", 6),
-                instanceURL: this.connection.url,
-                apiKey: this.connection.apiKey,
-                logging: this.logger,
-            };
-            const client = await Client.create(config);
+            const client = await getClient(this.connection, this.logger, "runbook", "run", 6);
 
             createRunbookRunFromInputs(client, command, this.task, this.logger);
 

--- a/source/tasks/Utils/client.ts
+++ b/source/tasks/Utils/client.ts
@@ -1,0 +1,18 @@
+import { Client, ClientConfiguration, Logger } from "@octopusdeploy/api-client";
+import { getUserAgentApp } from "./pluginInformation";
+import https from "https";
+import { OctoServerConnectionDetails } from "./connection";
+
+export async function getClient(connection: OctoServerConnectionDetails, logger: Logger, stepNoun: string, stepVerb: string, stepVersion: number) {
+    const config: ClientConfiguration = {
+        userAgentApp: getUserAgentApp(stepNoun, stepVerb, stepVersion),
+        instanceURL: connection.url,
+        apiKey: connection.apiKey,
+        logging: logger,
+    };
+    if (connection.ignoreSslErrors) {
+        config.httpsAgent = new https.Agent({ rejectUnauthorized: false });
+    }
+
+    return await Client.create(config);
+}

